src/mlmd/EdgeCanvas.tsx

```diff
@@ -23,83 +27,102 @@ interface EdgeCanvasProps {
-          data={[
-            {x: 0, y: y1},
-            {x: 30, y: y1},
-            {x: edgeWidth - 30, y: y4},
-            {x: edgeWidth, y: y4},
-          ]}
-          areaVisible={false}
-          axisVisible={false}
-          gridVisible={false}
-          labelsVisible={false}
-          pathColor={'#BDC1C6'}
-          pathVisible={true}
-          pathWidth={1}
-          pathOpacity={1}
-          pointsVisible={false}
-          viewBoxHeight={viewHeight}
-          viewBoxWidth={edgeWidth}
-          pathSmoothing={0}
-        />
-      );
-      viewHeight += cardBodyHeight;
-      lastNodePositions[lastNode] += cardBodyHeight;
-    });
-    viewHeight += cardOffset;
-    lastNodePositions[lastNode] += cardOffset;
-  });
+        const lastNode = reverseEdges ? 'y1' : 'y4';
+        const lastNodePositions = {y1: 0, y4: 0};
 
-  const edgeCanvasClasses = classes(css.edgeCanvas, reverseEdges ? css.edgeCanvasReverse: '');
-  return  <div className={edgeCanvasClasses}>{edgeLines}</div>
-};
+        const edgeLines: JSX.Element[] = [];
+        cards.forEach((card, i) => {
+            card.elements.forEach((element, j) => {
+                if (!element.next) return;
+                const {y1, y4} = lastNodePositions;
+                edgeLines.push(
+                    <LineChart
+                        key={`${i}-${j}`}
+                        data={[
+                            {x: 0, y: y1},
+                            {x: 30, y: y1},
+                            {x: viewWidth - 30, y: y4},
+                            {x: viewWidth     , y: y4},
+                        ]}
+                        areaVisible={false}
+                        axisVisible={false}
+                        gridVisible={false}
+                        labelsVisible={false}
+                        pathColor={'#BDC1C6'}
+                        pathVisible={true}
+                        pathWidth={1}
+                        pathOpacity={1}
+                        pointsVisible={false}
+                        viewBoxHeight={viewHeight}
+                        viewBoxWidth={viewWidth}
+                        pathSmoothing={0} />
+                );
+                viewHeight += cardBodyHeight;
+                lastNodePositions[lastNode] += cardBodyHeight;
+            });
+            viewHeight += cardOffset;
+            lastNodePositions[lastNode] += cardOffset;
+        });
+
+        const edgeCanvasClasses = classes(css.edgeCanvas, reverseEdges && 'edgeCanvasReverse');
+        sleep(50).then(this.componentDidMount.bind(this))
+        return <div className={edgeCanvasClasses}>{edgeLines}</div>
+    }
+
+    async componentDidMount() {
+        const $this = findDOMNode(this) as HTMLElement
+        const [$next, $parent] = [$this.nextSibling, $this.parentElement] as HTMLElement[]
+        if (!$this || !$next) return
+        await sleep(50) // So the browser has time to draw the elements (and have correct clientWidth values)
```

Comment:
Why 50? This sounds like something that could be brittle or different in other browsers or environments. Is there a way to make this value or logic more deterministic?

Reply:
Technically 1 should be good enough, 50 is being very generous (in case there are a ton of backlogged js async tasks the browser is trying to tackle (delayed scheduled callbacks on the event loop). I don't know of a more generic way